File: packages/capacitor-notificare/ios/Plugin/NotificarePlugin.swift
import Foundation
import Capacitor
import NotificareKit

@objc(NotificarePlugin)
public class NotificarePlugin: CAPPlugin {

    public override func load() {
        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Notificare.shared.delegate = self

        NotificationCenter.default.addObserver(self, selector: #selector(self.handleUrlOpened(notification:)), name: .capacitorOpenURL, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(self.handleUniversalLink(notification:)), name: .capacitorOpenUniversalLink, object: nil)
    }

    deinit {
        NotificationCenter.default.removeObserver(self, name: .capacitorOpenURL, object: nil)
        NotificationCenter.default.removeObserver(self, name: .capacitorOpenUniversalLink, object: nil)
    }

    @objc func handleUrlOpened(notification: NSNotification) {
        guard let object = notification.object as? [String: Any?], let url = object["url"] as? URL else {
            NotificareLogger.warning("Unprocessable url_opened event.")
            return
        }

        if Notificare.shared.handleTestDeviceUrl(url) {
            return
        }

        if Notificare.shared.handleDynamicLinkUrl(url) {
            return
        }

        EventBroker.instance.dispatchEvent("url_opened", data: ["url": url.absoluteString])
    }

    @objc func handleUniversalLink(notification: NSNotification) {
        guard let object = notification.object as? [String: Any?], let url = object["url"] as? URL else {
            NotificareLogger.warning("Unprocessable universal_link event.")
            return
        }

        if Notificare.shared.handleTestDeviceUrl(url) {
            return
        }

        _ = Notificare.shared.handleDynamicLinkUrl(url)
    }

    // MARK: - Notificare

    @objc func isConfigured(_ call: CAPPluginCall) {
        call.resolve([
            "result": Notificare.shared.isConfigured,
        ])
    }

    @objc func isReady(_ call: CAPPluginCall) {
        call.resolve([
            "result": Notificare.shared.isReady,
        ])
    }

    @objc func launch(_ call: CAPPluginCall) {
        onMainThread {
            Notificare.shared.launch()
            call.resolve()
        }
    }

    @objc func unlaunch(_ call: CAPPluginCall) {
        onMainThread {
            Notificare.shared.unlaunch()
            call.resolve()
        }
    }

    @objc func getApplication(_ call: CAPPluginCall) {
        do {
            var response: PluginCallResultData = [:]
            if let application = Notificare.shared.application {
                response["result"] = try application.toJson()
            }

            call.resolve(response)
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func fetchApplication(_ call: CAPPluginCall) {
        Notificare.shared.fetchApplication { result in
            switch result {
            case let .success(application):
                do {
                    call.resolve([
                        "result": try application.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func fetchNotification(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("Missing 'id' parameter.")
            return
        }

        Notificare.shared.fetchNotification(id) { result in
            switch result {
            case let .success(notification):
                do {
                    call.resolve([
                        "result": try notification.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func fetchDynamicLink(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("Missing 'url' parameter.")
            return
        }

        Notificare.shared.fetchDynamicLink(url) { result in
            switch result {
            case let .success(dynamicLink):
                do {
                    call.resolve([
                        "result": try dynamicLink.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    // MARK: - Notificare device module

    @objc func getCurrentDevice(_ call: CAPPluginCall) {
        do {
            var response: PluginCallResultData = [:]
            if let device = Notificare.shared.device().currentDevice {
                response["result"] = try device.toJson()
            }

            call.resolve(response)
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func getPreferredLanguage(_ call: CAPPluginCall) {
        var response: PluginCallResultData = [:]
        if let language = Notificare.shared.device().preferredLanguage {
            response["result"] = language
        }

        call.resolve(response)
    }

    @objc func updatePreferredLanguage(_ call: CAPPluginCall) {
        let language = call.getString("language")

        Notificare.shared.device().updatePreferredLanguage(language) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func register(_ call: CAPPluginCall) {
        let userId = call.getString("userId")
        let userName = call.getString("userName")

        onMainThread {
            Notificare.shared.device().register(userId: userId, userName: userName) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func fetchTags(_ call: CAPPluginCall) {
        Notificare.shared.device().fetchTags { result in
            switch result {
            case let .success(tags):
                call.resolve([
                    "result": tags
                ])
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func addTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Notificare.shared.device().addTag(tag) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func addTags(_ call: CAPPluginCall) {
        guard let tags = call.getArray("tags", String.self) else {
            call.reject("Missing 'tags' parameter.")
            return
        }

        Notificare.shared.device().addTags(tags) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func removeTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Notificare.shared.device().removeTag(tag) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func removeTags(_ call: CAPPluginCall) {
        guard let tags = call.getArray("tags", String.self) else {
            call.reject("Missing 'tags' parameter.")
            return
        }

        Notificare.shared.device().removeTags(tags) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func clearTags(_ call: CAPPluginCall) {
        Notificare.shared.device().clearTags { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func fetchDoNotDisturb(_ call: CAPPluginCall) {
        Notificare.shared.device().fetchDoNotDisturb { result in
            switch result {
            case let .success(dnd):
                do {
                    var response: PluginCallResultData = [:]
                    if let dnd = dnd {
                        response["result"] = try dnd.toJson()
                    }

                    call.resolve(response)
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func updateDoNotDisturb(_ call: CAPPluginCall) {
        guard let json = call.getObject("dnd") else {
            call.reject("Missing 'dnd' parameter.")
            return
        }

        let dnd: NotificareDoNotDisturb
        do {
            dnd = try NotificareDoNotDisturb.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        Notificare.shared.device().updateDoNotDisturb(dnd) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func clearDoNotDisturb(_ call: CAPPluginCall) {
        Notificare.shared.device().clearDoNotDisturb { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func fetchUserData(_ call: CAPPluginCall) {
        Notificare.shared.device().fetchUserData { result in
            switch result {
            case let .success(userData):
                call.resolve([
                    "result": userData
                ])
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func updateUserData(_ call: CAPPluginCall) {
        guard let json = call.getObject("userData") else {
            call.reject("Missing 'userData' parameter.")
            return
        }

        let userData = json.compactMapValues { $0 as? String }

        Notificare.shared.device().updateUserData(userData) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    // MARK: - Notificare events module

    @objc func logCustom(_ call: CAPPluginCall) {
        guard let event = call.getString("event") else {
            call.reject("Missing 'event' parameter.")
            return
        }

        let data = call.getObject("data")

        Notificare.shared.events().logCustom(event, data: data) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
}

extension NotificarePlugin: NotificareDelegate {
    public func notificare(_ notificare: Notificare, onReady application: NotificareApplication) {
        do {
            EventBroker.instance.dispatchEvent("ready", data: try application.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the ready event.", error: error)
        }
    }

    public func notificareDidUnlaunch(_ notificare: Notificare) {
        EventBroker.instance.dispatchEvent("unlaunched", data: nil)
    }

    public func notificare(_ notificare: Notificare, didRegisterDevice device: NotificareDevice) {
        do {
            EventBroker.instance.dispatchEvent("device_registered", data: try device.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the device_registered event.", error: error)
        }
    }
}

private func onMainThread(_ action: @escaping () -> Void) {
    DispatchQueue.main.async {
        action()
    }
}

File: packages/capacitor-notificare/android/src/main/java/re/notifica/capacitor/NotificarePlugin.kt
package re.notifica.capacitor

import android.content.Intent
import android.net.Uri
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONObject
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.ktx.device
import re.notifica.ktx.events
import re.notifica.models.*

@CapacitorPlugin(name = "NotificarePlugin")
public class NotificarePlugin : Plugin() {
    override fun load() {
        EventBroker.setup(this::notifyListeners)
        Notificare.intentReceiver = NotificarePluginIntentReceiver::class.java

        val intent = activity?.intent
        if (intent != null) handleOnNewIntent(intent)
    }

    override fun handleOnNewIntent(intent: Intent) {
        // Try handling the test device intent.
        if (Notificare.handleTestDeviceIntent(intent)) return

        // Try handling the dynamic link intent.
        val activity = activity
        if (activity != null && Notificare.handleDynamicLinkIntent(activity, intent)) return

        val url = intent.data?.toString()
        if (url != null) {
            EventBroker.dispatchEvent("url_opened", JSObject().apply {
                put("url", url)
            })
        }
    }

    // region Notificare

    @PluginMethod
    public fun isConfigured(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.isConfigured)
            }
        )
    }

    @PluginMethod
    public fun isReady(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.isReady)
            }
        )
    }

    @PluginMethod
    public fun launch(call: PluginCall) {
        Notificare.launch()
        call.resolve()
    }

    @PluginMethod
    public fun unlaunch(call: PluginCall) {
        Notificare.unlaunch()
        call.resolve()
    }

    @PluginMethod
    public fun getApplication(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.application?.toJson())
            }
        )
    }

    @PluginMethod
    public fun fetchApplication(call: PluginCall) {
        Notificare.fetchApplication(object : NotificareCallback<NotificareApplication> {
            override fun onSuccess(result: NotificareApplication) {
                call.resolve(
                    JSObject().apply {
                        put("result", result.toJson())
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchNotification(call: PluginCall) {
        val id = call.getString("id") ?: run {
            call.reject("Missing 'id' parameter.")
            return
        }

        Notificare.fetchNotification(id, object : NotificareCallback<NotificareNotification> {
            override fun onSuccess(result: NotificareNotification) {
                call.resolve(
                    JSObject().apply {
                        put("result", result.toJson())
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchDynamicLink(call: PluginCall) {
        val url = call.getString("url") ?: run {
            call.reject("Missing 'url' parameter.")
            return
        }

        val uri = Uri.parse(url)

        Notificare.fetchDynamicLink(uri, object : NotificareCallback<NotificareDynamicLink> {
            override fun onSuccess(result: NotificareDynamicLink) {
                call.resolve(
                    JSObject().apply {
                        put("result", result.toJson())
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun canEvaluateDeferredLink(call: PluginCall) {
        Notificare.canEvaluateDeferredLink(object : NotificareCallback<Boolean> {
            override fun onSuccess(result: Boolean) {
                call.resolve(
                    JSObject().apply {
                        put("result", result)
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun evaluateDeferredLink(call: PluginCall) {
        Notificare.evaluateDeferredLink(object : NotificareCallback<Boolean> {
            override fun onSuccess(result: Boolean) {
                call.resolve(
                    JSObject().apply {
                        put("result", result)
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // endregion

    // region Notificare device module

    @PluginMethod
    public fun getCurrentDevice(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.device().currentDevice?.toJson())
            }
        )
    }

    @PluginMethod
    public fun getPreferredLanguage(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.device().preferredLanguage)
            }
        )
    }

    @PluginMethod
    public fun updatePreferredLanguage(call: PluginCall) {
        val language = call.getString("language")

        Notificare.device().updatePreferredLanguage(language, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun register(call: PluginCall) {
        val userId = call.getString("userId")
        val userName = call.getString("userName")

        Notificare.device().register(userId, userName, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchTags(call: PluginCall) {
        Notificare.device().fetchTags(object : NotificareCallback<List<String>> {
            override fun onSuccess(result: List<String>) {
                call.resolve(
                    JSObject().apply {
                        put("result", JSArray(result))
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun addTag(call: PluginCall) {
        val tag = call.getString("tag") ?: run {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Notificare.device().addTag(tag, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun addTags(call: PluginCall) {
        val tags = call.getArray("tags", null)?.toList<String>() ?: run {
            call.reject("Missing 'tags' parameter.")
            return
        }

        Notificare.device().addTags(tags, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun removeTag(call: PluginCall) {
        val tag = call.getString("tag") ?: run {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Notificare.device().removeTag(tag, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun removeTags(call: PluginCall) {
        val tags = call.getArray("tags", null)?.toList<String>() ?: run {
            call.reject("Missing 'tags' parameter.")
            return
        }

        Notificare.device().removeTags(tags, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun clearTags(call: PluginCall) {
        Notificare.device().clearTags(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchDoNotDisturb(call: PluginCall) {
        Notificare.device().fetchDoNotDisturb(object : NotificareCallback<NotificareDoNotDisturb?> {
            override fun onSuccess(result: NotificareDoNotDisturb?) {
                call.resolve(
                    JSObject().apply {
                        put("result", result?.toJson())
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun updateDoNotDisturb(call: PluginCall) {
        val dnd = call.getObject("dnd", null)?.let { NotificareDoNotDisturb.fromJson(it) } ?: run {
            call.reject("Missing 'dnd' parameter.")
            return
        }

        Notificare.device().updateDoNotDisturb(dnd, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun clearDoNotDisturb(call: PluginCall) {
        Notificare.device().clearDoNotDisturb(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchUserData(call: PluginCall) {
        Notificare.device().fetchUserData(object : NotificareCallback<NotificareUserData> {
            override fun onSuccess(result: NotificareUserData) {
                val userData = JSONObject().apply {
                    result.forEach {
                        put(it.key, it.value)
                    }
                }

                call.resolve(
                    JSObject().apply {
                        put("result", userData)
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun updateUserData(call: PluginCall) {
        val json = call.getObject("userData", null) ?: run {
            call.reject("Missing 'userData' parameter.")
            return
        }

        val userData: NotificareUserData = mutableMapOf<String, String>().apply {
            val iterator = json.keys()
            while (iterator.hasNext()) {
                val key = iterator.next()
                val value = json.getString(key)

                if (value != null) {
                    put(key, value)
                }
            }
        }

        Notificare.device().updateUserData(userData, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // endregion

    // region Notificare events module

    @PluginMethod
    public fun logCustom(call: PluginCall) {
        val event = call.getString("event") ?: run {
            call.reject("Missing 'event' parameter.")
            return
        }

        val data: NotificareEventData?

        try {
            data = call.getObject("data", null)?.let { NotificareEvent.createData(it) }
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.events().logCustom(event, data, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // endregion
}
